addons/hr_attendance/models/hr_attendance.py

```diff
@@ -99,6 +104,194 @@ def _check_validity(self):
                         'datetime': format_datetime(self.env, last_attendance_before_check_out.check_in, dt_format=False),
                     })
 
+    def _get_tz_cache(self):
+        #Create a cache for the different timedelta for the timezones we will encounter
+        #Adding the delta to a naive date is the same as localizing and converting timezone but less costly
+        return {tz: datetime.now(pytz.timezone(tz)).utcoffset() for tz in set(self.employee_id._get_tz().values()) | {'UTC'}}
+
+    def _get_attendances_dates(self, tz_delta_cache):
+        # Returns a dictionnary {employee_id: set(dates)}
+        attendances_emp = defaultdict(set)
+        for attendance in self.filtered(lambda a: a.employee_id.company_id.hr_attendance_overtime and a.check_in):
+            tz = attendance.employee_id.tz or 'UTC'
+            check_in = attendance.check_in + tz_delta_cache[tz]
```

Comment:
So we'll deal with naïve datetime in a different timezone than UTC, right. It looks really weird to me.

Reply:
The handling of dates is essentially the same as before, we just avoid rewriting the whole `localize`, `astimezone` and `replace(tzinfo..)` every time, if we could query using timezones directly this would be easily avoided but unfortunately that's not the case ..

As for `Second, we have many calls to datetime.date(), which means we delegate the responsibility to the python library to return the date of a naive datetime.`
I am not quite sure what the issue would be with the date() function calls, could you elaborate? At first glance it seems like a really simple function here: https://github.com/python/cpython/blob/main/Lib/datetime.py#L1810-L1812

I will take another fresh look at all of this I am sure we can find a solution that will please both of us 😄

Reply:
The problem I noticed was (related to the fact we deal with naive non-UTC datetime) : You want the day of a naive datetime, so we know which day we are, but we don't know where on earth (we have to read the whole code to understand the "where")?

So I think what you proposed at last is a solution more clear and understandable, we don't have to read the whole code to understand what we are doing with the datetime/date/timezones. (& it makes me happy :) )